test/docs/jdk/javadoc/doccheck/doccheckutils/checkers/LinkChecker.java

```diff
@@ -0,0 +1,468 @@
+                    p = currFile;
+                } else {
+                    p = currFile.getParent().resolve(resolvedUriPath).normalize();
+                }
+
+                if (fragment != null && !fragment.isEmpty()) {
+                    foundReference(line, p, fragment);
+                }
+            }
+        } catch (URISyntaxException e) {
+            System.err.println("Failed to create URI: " + ref);
+            log.log(currFile, line, "invalid URI: " + e);
+        }
+    }
+
+
+    private void foundReference(int line, Path p, String fragment) {
+        IDTable t = allFiles.computeIfAbsent(p, key -> new IDTable(log.relativize(key)));
+        t.addReference(fragment, currFile, line);
+    }
+
+    private void foundReference(int line, URI uri) {
+        if (!isSchemeOK(uri.getScheme()) && !checkInwardReferencesOnly) {
+            log.log(currFile, line, "bad scheme in URI");
+            badSchemes++;
+        }
+
+        String fragment = uri.getRawFragment();
+        if (fragment != null && !fragment.isEmpty()) {
+            try {
+                URI noFrag = new URI(uri.toString().replaceAll("#\\Q" + fragment + "\\E$", ""));
+                IDTable t = allURIs.computeIfAbsent(noFrag, IDTable::new);
+                t.addReference(fragment, currFile, line);
+            } catch (URISyntaxException e) {
+                throw new Error(e);
+            }
+        }
+    }
+
+    private boolean isSchemeOK(String uriScheme) {
+        if (uriScheme == null) {
+            return true;
+        }
+
+        return switch (uriScheme) {
+            case "ftp", "http", "https", "javascript" -> true;
+            default -> false;
+        };
+    }
+
+    static class Position implements Comparable<Position> {
+        Path path;
+        int line;
+
+        Position(Path path, int line) {
+            this.path = path;
+            this.line = line;
+        }
+
+        @Override
+        public int compareTo(Position o) {
+            int v = path.compareTo(o.path);
+            return v != 0 ? v : Integer.compare(line, o.line);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            } else if (obj == null || getClass() != obj.getClass()) {
+                return false;
+            } else {
+                final Position other = (Position) obj;
+                return Objects.equals(this.path, other.path)
+                        && this.line == other.line;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(path) * 37 + line;
+        }
+    }
+
+    static class IDInfo {
+        boolean declared;
+        Set<Position> references;
+
+        Set<Position> getReferences() {
+            return references == null ? Collections.emptySet() : references;
+        }
+    }
+
+    static class URIComparator implements Comparator<URI> {
```

Comment:
It looks like this class is unused.

Reply:
Will remove it, it could be used to print out links in the same order everytime but it doesn't seem important now.